src/group/mod.rs

```diff
@@ -157,31 +153,51 @@ impl<'a> Groups {
 
     /// Update given `peer`'s group ID. Return `true` if any peers are unresolved.
     pub(crate) fn update_group_id(&mut self, peer: u64, group_id: u64) -> bool {
-        let mut remove_delegate = false;
-        match self.indexes.entry(peer) {
-            MapEntry::Occupied(e) => {
-                if group_id == INVALID_ID {
-                    let (_, (_, d)) = e.remove_entry();
-                    remove_delegate = d == peer;
-                } else {
-                    let (ref mut gid, ref mut d) = e.into_mut();
-                    if *gid != group_id {
-                        *gid = group_id;
-                        *d = INVALID_ID;
-                        self.unresolved.push(peer);
-                        remove_delegate = *d == peer;
-                    }
-                }
+        if group_id == INVALID_ID {
+            self.unmark_peer(peer);
+        } else if let Some((gid, _)) = self.indexes.get(&peer) {
+            if *gid == group_id {
+                return false;
+            }
+            self.unmark_peer(peer);
+            self.mark_peer(peer, group_id);
+        } else {
+            self.mark_peer(peer, group_id);
+        }
+        !self.unresolved.is_empty()
+    }
+
+    fn unmark_peer(&mut self, peer: u64) {
```

Comment:
Does this denote that `peer` is removed from the `Groups` system?

Reply:
Yes.